File: cppStudy/solved/1865 웜홀.cpp
#include <bits/stdc++.h>
using namespace std;
// 1865 웜홀
#define FOR(i, n, m) for(int i=(n); i<(m); i++)
#define cost first
#define node second
const int VLIM = 502;
const int INF = 1000000007;
using TI3 = tuple<int, int, int>;
using PII = pair<int, int>;

int V, R, W, a, b, c;
bool minusLoop(vector<PII> adj[]) {
	vector<int> upper(V+1, INF);
	bool updated = false;
    upper[1] = 0;
	FOR(i, 0, V) {
		updated = false;
		FOR(v, 1, V+1) {
			for (auto nv : adj[v]) {
				if (upper[v] + nv.cost < upper[nv.node]) {
					updated = true;
					upper[nv.node] = upper[v] + nv.cost;
				}
			}
		}
		if (!updated) break;
	}
	return updated;
}
int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	int tc;
	cin >> tc;
	FOR(ttt, 0, tc) {
		vector<PII> adj[VLIM] = {};
		
		cin >> V >> R >> W;
		FOR(i, 0, R) {
			cin >> a >> b >> c;
			adj[a].push_back({ c, b });
			adj[b].push_back({ c, a });
		}
		FOR(i, 0, W) {
			cin >> a >> b >> c;
			adj[a].push_back({ -c, b });
		}
		if (minusLoop(adj)) cout << "YES" << '\n';
		else cout << "NO" << '\n';
	}
}

File: pythonStudy/BOJ/1865 웜홀.py
# 1865 웜홀
input = __import__('sys').stdin.readline
INF = 10e9+1
for _ in range(int(input())):
    n, r, w = map(int, input().split())
    adj = [[] for _ in range(n+1)]
    for i in range(r):
        a, b, c = map(int, input().split())
        adj[a].append((c, b))
        adj[b].append((c, a))
    for i in range(w):
        a, b, c = map(int, input().split())
        adj[a].append((-c, b))

    upper = [INF]*(n+1)
    upper[1] = 0
    updated = False
    for i in range(n):
        updated = False
        for v in range(1, n+1):
            for nv in adj[v]:
                if upper[v] + nv[0] < upper[nv[1]]:
                    updated = True
                    upper[nv[1]] = upper[v] + nv[0]
        if not updated:
            print('NO')
            break
    if updated:
        print('YES')